tests/test_update_query.py

```diff
@@ -117,6 +117,43 @@ def test_with_query_list_int():
     assert str(url.with_query([("a", 1)])) == "http://example.com/?a=1"
 
 
+@pytest.mark.parametrize(
+    "query,expected",
+    [
+        pytest.param({"a": []}, "?", id="empty list"),
+        pytest.param({"a": ()}, "?", id="empty tuple"),
+        pytest.param({"a": [1]}, "?a=1", id="single list"),
+        pytest.param({"a": (1,)}, "?a=1", id="single tuple"),
+        pytest.param({"a": [1, 2]}, "?a=1&a=2", id="list"),
+        pytest.param({"a": (1, 2)}, "?a=1&a=2", id="tuple"),
+        pytest.param({"a[]": [1, 2]}, "?a[]=1&a[]=2", id="key with braces"),
+        pytest.param({"&": [1, 2]}, "?%26=1&%26=2", id="quote key"),
+        pytest.param({"a": ["1", 2]}, "?a=1&a=2", id="mixed types"),
+        pytest.param({"&": ["=", 2]}, "?%26=%3D&%26=2", id="quote key and value"),
+        pytest.param({"a": 1, "b": [2, 3]}, "?a=1&b=2&b=3", id="single then list"),
+        pytest.param({"a": [1, 2], "b": 3}, "?a=1&a=2&b=3", id="list then single"),
+        pytest.param({"a": ["1&a=2", 3]}, "?a=1%26a%3D2&a=3", id="ampersand then int"),
+        pytest.param({"a": [1, "2&a=3"]}, "?a=1&a=2%26a%3D3", id="int then ampersand"),
+    ],
+)
+def test_with_query_sequence(query, expected):
+    url = URL("http://example.com")
+    assert url.with_query(query) == URL("http://example.com/" + expected)
```

Comment:
Maybe it'll be better to compare strings? Currently you're testing that the main initializer does the same thing as `URL.with_query()` method.
```suggestion
    assert str(url.with_query(query)) == "http://example.com/{expected}".format_map(locals())
```